File: recover/src/recover/jobs/outbox.py
from __future__ import annotations

from sqlalchemy.ext.asyncio import AsyncEngine

from recover import metrics, repo
from recover.config import JobConfig


class OutboxRecoveryJob:
    name = "outbox_recovery"

    def __init__(self, engine: AsyncEngine, cfg: JobConfig, regions: list[str]) -> None:
        self.engine = engine
        self.cfg = cfg
        self.regions = regions

    async def tick(self) -> int:
        n = await repo.reset_stuck_outbox(self.engine, self.cfg.batch_size, self.regions)
        if n > 0:
            metrics.outbox_unstuck_total.inc(n)
        return n

File: recover/src/jobs/outbox.rs
use anyhow::Result;
use async_trait::async_trait;

use crate::config::JobConfig;
use crate::jobs::{Job, JobOutcome, ServiceContext};
use crate::repo;

pub struct OutboxRecovery {
    pub cfg: JobConfig,
}

#[async_trait]
impl Job for OutboxRecovery {
    fn name(&self) -> &'static str {
        "outbox_recovery"
    }

    fn config(&self) -> &JobConfig {
        &self.cfg
    }

    async fn tick(&self, ctx: &ServiceContext) -> Result<JobOutcome> {
        let rows = repo::reset_stuck_outbox(
            &ctx.pool,
            self.cfg.batch_size as i32,
            ctx.regions(),
        )
        .await?;
        if rows > 0 {
            tracing::info!(rows, "outbox events unstuck");
            metrics::counter!("recover_outbox_unstuck_total").increment(rows);
            Ok(JobOutcome::Processed(rows))
        } else {
            Ok(JobOutcome::Idle)
        }
    }
}
